### ISSUE TITLE
Training Discussion 

### ISSUE THREAD

**Issue by kdcyberdude** [0417keito/VALL-E-S#1]
Thank you @0417keito, for providing the trainer code for VALL-E-X. I am currently in the process of modifying the Auto Regressive model code to dub audio into a different language. 

To illustrate, the input would be _English_Output_Text + Hindi_Audio_Prefix_, and the output would be _English_Audio_.  Where I want the number of quantized frames of _English_Audio_ to match _Hindi_Audio_Prefix_. 

I have made the following changes - Here **en_audio_features** is English_Audio and **hi_audio_features** is Hindi_Audio_Prefix.
The fundamental idea in the following code is that, I have utilized the combined text and Hindi_Audio_prefix in **'x'** and generated sequence audio i.e. English_Audio as **'y'**. Accordingly, I have adjusted the attention mask.
```python
# NOTE: x has been padded in TextTokenCollater
        x_mask = make_pad_mask(x_lens).to(x.device)
        y_mask = make_pad_mask(y_lens).to(en_audio_features.device)
        y_mask_int = y_mask.type(torch.int64)

        text = x
        codes = en_audio_features.type(torch.int64) * (1 - y_mask_int.unsqueeze(dim=-1))
        codes_prefix = hi_audio_features.type(torch.int64) * (1 - y_mask_int.unsqueeze(dim=-1))

        y, targets = self.pad_y_eos(
            codes[..., 0], y_mask_int, eos_id=NUM_AUDIO_TOKENS
        )

        y_prefix, targets_prefix = self.pad_y_eos(
            codes_prefix[..., 0], y_mask_int, eos_id=NUM_AUDIO_TOKENS
        )

        x_len = x_lens.max()

        metrics = {}
        total_loss = 0.0

        xy_padding_mask = torch.concat([x_mask, y_mask, y_mask], dim=1)
        if self.ar_audio_prepend_bos:
            ar_xy_padding_mask = torch.concat(
                [x_mask, F.pad(y_mask, (1, 0), value=False), F.pad(y_mask, (1, 0), value=False)], dim=1
            )
        else:
            ar_xy_padding_mask = xy_padding_mask
        # AR Decoder
        if train_stage in [0, 1]:
            x = self.ar_text_embedding(text)
            x = self.ar_text_prenet(x)
            x = self.ar_text_position(x)

            y_len = y_lens.max() + int(self.ar_audio_prepend_bos)

            x_attn_mask = F.pad(
                torch.zeros((x_len + y_len, x_len + y_len), dtype=torch.bool, device=x.device),
                (0, y_len),
                value=True,
            )

            y_attn_mask = F.pad(
                torch.triu(
                    torch.ones(y_len, y_len, dtype=torch.bool, device=x.device),
                    diagonal=1,
                ),
                (x_len + y_len, 0),
                value=False,
            )
            xy_attn_mask = torch.concat([x_attn_mask, y_attn_mask], dim=0)

            # merge key padding and attention masks
            bsz, src_len = x.shape[0], x_len + y_len + y_len
            _xy_padding_mask = (
                ar_xy_padding_mask.view(bsz, 1, 1, src_len)
                .expand(-1, self.num_heads, -1, -1)
                .reshape(bsz * self.num_heads, 1, src_len)
            )
            xy_attn_mask = xy_attn_mask.logical_or(_xy_padding_mask)

            new_attn_mask = torch.zeros_like(xy_attn_mask, dtype=x.dtype)
            new_attn_mask.masked_fill_(xy_attn_mask, float("-inf"))
            xy_attn_mask = new_attn_mask

            y = torch.concat([y_prefix, y], dim=1)

            y_emb = self.ar_audio_embedding(y)
            y_emb = self.ar_audio_prenet(y_emb)
            y_pos = self.ar_audio_position(y_emb)

            xy_pos = torch.concat([x, y_pos], dim=1)

            xy_dec, _ = self.ar_decoder(
                (xy_pos, None),
                mask=xy_attn_mask,
                # src_key_padding_mask=xy_padding_mask,
                # is_causal=True,
            )
            logits = self.ar_predict_layer(xy_dec[:, x_len + y_len:]).permute(0, 2, 1)
            # loss
            total_loss = F.cross_entropy(logits, targets, reduction=reduction)

            metrics["ArTop10Accuracy"] = self.ar_accuracy_metric(
                logits.detach(), targets
            ).item() * y_lens.sum().type(torch.float32)

        if self.num_quantizers == 1:
            return ((x, codes), total_loss, metrics)

        # Non-AR Decoders
```

I am eager to discuss and brainstorm about the approach. Does the aforementioned method seem viable, or should we consider an alternative strategy? Since I am new to this I'll really appreciate the help!!
